rust/datafusion/src/execution/physical_plan/expressions.rs

```diff
@@ -199,6 +200,66 @@ pub fn sum(expr: Arc<dyn PhysicalExpr>) -> Arc<dyn AggregateExpr> {
     Arc::new(Sum::new(expr))
 }
 
+/// COUNT aggregate expression
+/// If given something like `SELECT COUNT(*) FROM table`, the number of rows in the table will be returned.
+/// If given a specific column, `SELECT COUNT(column_name) FROM table`, the number of *non-null*
+/// rows in that column will be returned.
+pub struct Count {
+    expr: Arc<dyn PhysicalExpr>,
+}
+
+impl Count {
+    /// Create a new COUNT aggregate function.
+    pub fn new(expr: Arc<dyn PhysicalExpr>) -> Self {
+        Self { expr: expr }
+    }
+}
+
+impl AggregateExpr for Count {
+    fn name(&self) -> String {
+        "COUNT".to_string()
+    }
+
+    fn data_type(&self, _input_schema: &Schema) -> Result<DataType> {
+        Ok(DataType::UInt64)
+    }
+
+    fn create_accumulator(&self) -> Rc<RefCell<dyn Accumulator>> {
+        Rc::new(RefCell::new(CountAccumulator {
+            expr: self.expr.clone(),
+            count: 0,
+        }))
+    }
+
+    fn create_combiner(&self, column_index: usize) -> Arc<dyn AggregateExpr> {
+        Arc::new(Sum::new(Arc::new(Column::new(column_index))))
+    }
+}
+
+struct CountAccumulator {
+    expr: Arc<dyn PhysicalExpr>,
+    count: u64,
+}
+
+impl Accumulator for CountAccumulator {
+    fn accumulate(&mut self, batch: &RecordBatch, row_index: usize) -> Result<()> {
+        let array = self.expr.evaluate(batch)?;
```

Comment:
I just spotted an issue with this, and I have the same issue with the `SumExpr` implementation ... we are evaluating the expression against the whole batch multiple times (once for every row in the batch). This is a design flaw in the accumulator trait I guess. I'll give this some thought today.

Reply:
I think it would be best if we merge this in without this optimization/fix, so you can simply fix both instances at the same time?

Reply:
Please take a look at the proposed fix in https://github.com/apache/arrow/pull/5542 and let me know what you think. I'd prefer to get this reviewed and merged first, then you can rebase this PR and implement the changes.